pulsar-testclient/src/main/java/com/yahoo/pulsar/testclient/LoadSimulationServer.java

```diff
@@ -0,0 +1,357 @@
+	private static class TradeUnit {
+		Future<Producer> producerFuture;
+		Future<Consumer> consumerFuture;
+		final AtomicBoolean stop;
+		final RateLimiter rateLimiter;
+
+		// Creating a byte[] for every message is stressful for a client
+		// machine, so in order to ensure that any
+		// message size may be sent/changed while reducing object creation, the
+		// byte[] is wrapped in an AtomicReference.
+		final AtomicReference<byte[]> payload;
+		final ProducerConfiguration producerConf;
+		final PulsarClient client;
+		final String topic;
+		final Map<Integer, byte[]> payloadCache;
+
+		public TradeUnit(final TradeConfiguration tradeConf, final PulsarClient client,
+				final ProducerConfiguration producerConf, final ConsumerConfiguration consumerConf,
+				final Map<Integer, byte[]> payloadCache) throws Exception {
+			consumerFuture = client.subscribeAsync(tradeConf.topic, "Subscriber-" + tradeConf.topic, consumerConf);
+			producerFuture = client.createProducerAsync(tradeConf.topic, producerConf);
+			this.payload = new AtomicReference<>();
+			this.producerConf = producerConf;
+			this.payloadCache = payloadCache;
+			this.client = client;
+			topic = tradeConf.topic;
+
+			// Add a byte[] of the appropriate size if it is not already present
+			// in the cache.
+			this.payload.set(payloadCache.computeIfAbsent(tradeConf.size, byte[]::new));
+			rateLimiter = RateLimiter.create(tradeConf.rate);
+			stop = new AtomicBoolean(false);
+		}
+
+		// Change the message rate/size according to the given configuration.
+		public void change(final TradeConfiguration tradeConf) {
+			rateLimiter.setRate(tradeConf.rate);
+			this.payload.set(payloadCache.computeIfAbsent(tradeConf.size, byte[]::new));
+		}
+
+		// Attempt to create a Producer indefinitely. Useful for ensuring
+		// messages continue to be sent after broker
+		// restarts occur.
+		private Producer getNewProducer() throws Exception {
+			while (true) {
+				try {
+					return client.createProducerAsync(topic, producerConf).get();
+				} catch (Exception e) {
+					Thread.sleep(10000);
+				}
+			}
+		}
+
+		private class MutableBoolean {
+			public volatile boolean value = true;
+		}
+
+		public void start() throws Exception {
+			Producer producer = producerFuture.get();
+			final Consumer consumer = consumerFuture.get();
+			while (!stop.get()) {
+				final MutableBoolean wellnessFlag = new MutableBoolean();
+				final Function<Throwable, ? extends MessageId> exceptionHandler = e -> {
+					// Unset the well flag in the case of an exception so we can
+					// try to get a new Producer.
+					wellnessFlag.value = false;
+					return null;
+				};
+				while (!stop.get() && wellnessFlag.value) {
+					producer.sendAsync(payload.get()).exceptionally(exceptionHandler);
+					rateLimiter.acquire();
+				}
+				producer.closeAsync();
+				if (!stop.get()) {
+					// The Producer failed due to an exception: attempt to get
+					// another producer.
+					producer = getNewProducer();
```

Comment:
Hmm, it seems that for whatever reason, I am experiencing issues related to netty/direct memory if I do not attempt to make a new producer:
```
io.netty.util.IllegalReferenceCountException: refCnt: 0, increment: 1
        at io.netty.buffer.AbstractReferenceCountedByteBuf.retain(AbstractReferenceCountedByteBuf.java:63)
        at com.yahoo.pulsar.common.compression.CompressionCodecNone.encode(CompressionCodecNone.java:27)
        at com.yahoo.pulsar.client.impl.BatchMessageContainer.getCompressedBatchMetadataAndPayload(BatchMessageContainer.java:107)
        at com.yahoo.pulsar.client.impl.ProducerImpl.batchMessageAndSend(ProducerImpl.java:1019)
        at com.yahoo.pulsar.client.impl.ProducerImpl.doBatchSendAndAdd(ProducerImpl.java:293)
        at com.yahoo.pulsar.client.impl.ProducerImpl.sendAsync(ProducerImpl.java:237)
        at com.yahoo.pulsar.client.impl.ProducerImpl.sendAsync(ProducerImpl.java:134)
        at com.yahoo.pulsar.client.impl.ProducerBase.sendAsync(ProducerBase.java:47)
        at com.yahoo.pulsar.testclient.LoadSimulationServer$TradeUnit.start(LoadSimulationServer.java:112)
        at com.yahoo.pulsar.testclient.LoadSimulationServer.lambda$handle$5(LoadSimulationServer.java:212)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
        at java.lang.Thread.run(Thread.java:745)
```
As a result, I can end up losing a bunch of bundles in the simulation. I will leave the get new producer logic for now and try to find a better solution later